src/dns.c

```diff
@@ -2527,119 +2545,123 @@ hsk_dns_name_cmp(const char *a, const char *b) {
   if (!b)
     return 1;
 
-  size_t alen = strlen(a);
-  size_t blen = strlen(b);
+  uint8_t off = 0;
 
-  if (alen > 0 && a[alen - 1] == '.')
-    alen -= 1;
+  for (;;) {
+    // compare label length bytes
+    uint8_t labela = a[off];
+    uint8_t labelb = b[off];
 
-  if (blen > 0 && b[blen - 1] == '.')
-    blen -= 1;
+    if (labela < labelb)
+      return -1;
 
-  size_t len = alen < blen ? alen : blen;
+    if (labela > labelb)
+      return 1;
 
-  int i;
-  for (i = 0; i < len; i++) {
-    uint8_t x = (uint8_t)a[i];
-    uint8_t y = (uint8_t)b[i];
+    if (labela == 0)
+      return 0;
 
-    if (x >= 0x41 && x <= 0x5a)
-      x |= 0x20;
+    // labels are equal non-zero length: compare label
+    for (; labela > 0; labela--) {
+      off++;
 
-    if (y >= 0x41 && y <= 0x5a)
-      y |= 0x20;
+      if (a[off] > b[off])
+        return 1;
 
-    if (x < y)
-      return -1;
+      if (a[off] < b[off])
+        return -1;
+    }
 
-    if (x > y)
-      return 1;
+    off++;
   }
 
-  if (alen < blen)
-    return -1;
+  return 0;
+}
 
-  if (alen > blen)
-    return 1;
+bool
+hsk_dns_name_equal(const void *a, const void *b) {
```

Comment:
I think it's because we pass this function to `hsk_map_init_map()` which accepts as an argument one of these:

```
typedef bool (*hsk_map_equal_func)(const void *a, const void *b);
```